docs/agent/config.md

```diff
@@ -87,7 +87,191 @@ differently from Agent version 5.
+This handle can be used as a *value* of any YAML fields in your configuration
+(not the key) in any section (`init_config`, `instances`, `logs`, ...).
+
+Encrypted secrets are supported in every configuration backend: file, etcd,
+consul ...
+
+Example:
+
+```yaml
+instances:
+  - server: db_prod
+    user: ENC[db_prod_user]
+    password: ENC[db_prod_password]
+```
+
+In the above example we have two encrypted  secrets : `db_prod_user` and
+`db_prod_password`. Those are the secrets **handle** and must uniquely identify
+a secret within your secrets management tool.
+
+Between the brackets every character is allowed as long as the configuration
+YAML is valid. This means you could use any format you want.
+
+Example:
+
+```
+ENC[{"env": "prod", "check": "postgres", "id": "user_password", "az": "us-east-1a"}]
+```
+
+In this example the secret handle is the string `{"env": "prod", "check":
+"postgres", "id": "user_password", "az": "us-east-1a"}`.
+
+**Autodiscovery**:
+
+Secrets are resolved **after** Autodiscovery template variables. This means you
+can use them in a secret handle to fetch secrets specific to a container.
+
+Example:
+
+```
+instances:
+  - server: %%host%%
+    user: ENC[db_prod_user_%%host%%]
+    password: ENC[db_prod_password_%%host%%]
+```
+
+### Fetching secrets
+
+To fetch the secrets from you configurations you have to provide an executable
+capable of fetching passwords from your secrets management tool.
+
+An external executable solved multiple issues:
+
+- Guarantees that the agent will never have more access/information than what
+  you transfer to it.
+- Maximum freedom and flexibility: allowing users to use any secrets management
+  tool (including closed sources ones) without having to rebuild the agent.
+- Solving the **initial trust** is a very complex problem as it usually changes
+  from users to users. Each setup requires more or less control and might rely on
+  private tools. This shifts the problem to where it is easier to solve, without
+  having to rebuild the agent.
+
+This executable will be executed by the agent once per check's instance. The
+agent will cache secrets internally to reduce the number of calls (useful in
+a containerized environment for example).
+
+The executable **MUST** (the agent will refuse to use it otherwise):
+
+- Belong to the same user running the agent (usually `dd-agent`).
+- Have **no** rights for `group` or `other`.
+- Have at least `exec` right for the owner.
+- The executable will not share any environment variables with the agent.
+
+#### Configuration
+
+In `datadog.yaml` you must set the following variables:
+
+```yaml
+secret_backend_command: /path/to/you/executable
+```
+
+More settings are available: see `datadog.yaml`.
+
+#### The executable API
+
+The executable has to respect a very simple API: it reads a JSON on the
+Standard input and output a JSON containing the decrypted secrets on the
+Standard output.
```

Comment:
only realizing this now, but it could be helpful for users to make the Agent log whatever the executable outputs on `stderr` (with a size limit if it makes sense security-wise). Allowing to use only the `error` fields in the json output for that could be too restrictive in case the executable encounters an unexpected error. Or was this intentionally not implemented for security reasons?

Reply:
Interesting point, we detect the exit code but don't handle stderr. I think we should. Should we plan this for the next version (card in the backlog) or do it right away ?